### PR TITLE
Add proposal for mixins

### PR DESCRIPTION
**Pull request by mtdowling**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: designs/mixins.md [hunk 1]
```diff
@@ -0,0 +1,1075 @@
...
+
+```
+structure C {
+    a: String,
+    b: String,
+    c: String
+}
+```
+
+Unions can be mixins and use mixins too.
```
Comment (JordonPhillips) [inline designs/mixins.md:167]: Can union mixins be used with structures and vice versa? e.g.

```
@mixin 
union UnionMixin {
    a: String
}

structure Struct with UnionMixin {
    b: String
}
```

Pretty sure the answer is no, but it should be said explicitly.

File: designs/mixins.md [hunk 2]
```diff
@@ -0,0 +1,1075 @@
...
+
+The `@mixin` trait has a single optional member named `localTraits` that
+contains a list of shape IDs. Each shape ID MUST reference a valid trait that
+is applied directly to the mixin. The traits referenced in `localTraits` are
+not copied onto shapes that use the mixin. `localTraits` only affects traits
+applied to the mixin container shape and has no impact on the members
+contained within a mixin.
+
+> Note: `smithy.api#mixin` is considered implicitly present in the
+> `localTraits` property and does not  need to be defined in the list of
```
Comment (JordonPhillips) [inline designs/mixins.md:770]:
```suggestion
> `localTraits` property and does not need to be defined in the list of
```

**Review by JordonPhillips — APPROVED** [review]

File: designs/mixins.md (hunk 1 above)
Comment (mtdowling) [reply designs/mixins.md:167]: No. There's a callout for this that states:
> Structure shapes can only use structure mixins, and union shapes can only use union mixins.

File: designs/mixins.md [hunk 3]
```diff
@@ -0,0 +1,1075 @@
...
+    input: InputMixin // <- this is invalid
+}
+```
+
+
+### Mixins in the IDL
+
+To support mixins, the `structure_statement` and `union_statement` ABNF rules
+will be updated to contain an optional `with` clause that comes after the
+shape name and before `{`. If present, the `with` clause MUST contain one or
```
Comment (joguSD) [inline designs/mixins.md:585]: How do `with` list members scale? Say I had a dozen or so mixins used on a particular shape, would they all be required to be defined in a single line? Or is there a way to breakup the definition like:

```
structure MyStruct with (
    MyMixinA,
    MyMixinB,
    MyMixinC,
...
) {
    ...
}
```

**Review by joguSD — APPROVED** [review]

File: designs/mixins.md (hunk 3 above)
Comment (mtdowling) [reply designs/mixins.md:585]: In the current proposal, whitespace and commas are insignificant, you can break them up onto different lines, and there's no need for a `(` or `)`:

```
structure MyStruct with
    MyMixinA
    MyMixinB
    MyMixinC
    MyMixinD
{
    @required
    foo: String
}
```

Comment (joguSD) [reply designs/mixins.md:585]: Cool, makes sense.

File: designs/mixins.md [hunk 4]
```diff
@@ -0,0 +1,1075 @@
...
+structure B with A {
+    /// B docs
+    @tags(["hello"])
+    @deprecated
+    b: String
+}
+
+structure ABC with A, B {
+    /// More specific B docs
+    @override
```
Comment (johnstonskj) [inline designs/mixins.md:369]: So the override here applies to the usage of the trait?

Comment (johnstonskj) [inline designs/mixins.md:369]: In this example we alter an existing trait, I can see how one would add one, but is it possible to remove a trait?

File: designs/mixins.md [hunk 5]
```diff
@@ -0,0 +1,1075 @@
...
+
+### Mixins and member ordering
+
+The ordering of structure and union members is important for languages like C
+that require a stable ABI. Mixins provide a deterministic member ordering.
+Members are ordered by:
+
+- Members copied from mixins come before members defined directly in the shape.
+- Members are copied from mixins based on the order in which they applied to
+  a shape.
```
Comment (johnstonskj) [inline designs/mixins.md:706]: Is it the order in which they were _first_, or _last_ applied to the shape?

File: designs/mixins.md [hunk 6]
```diff
@@ -0,0 +1,1075 @@
...
+
+```
+@trait(selector: "is(structure, union) > member")
+structure override {}
+```
+
+
+### Reference implementation notes
+
+Mixins will be implemented in Smithy's Java reference implementation in a way
```
Comment (johnstonskj) [inline designs/mixins.md:786]: Does this imply that other implementations could choose a different behavior, or is the reference implementation the preferred approach?

**Comment by johnstonskj** [comment]
Will this imply a new version number? it seems like we introduced a compatibility issue when removing commas without a version bump. 

As for the BNF, is this a reasonable approximation?

```ebnf
structure_statement =
    "structure" ws identifier ws [mixins ws] shape_members

mixins =
    "with" ws root_shape_id *(ws comma root_shape_id) trailing_comma
```

**Comment by mtdowling** [comment]
@johnstonskj the removal of commas was never released. It's only in the idl-1.1 branch. This proposal would be built on top of that branch and eventually released with a new IDL version of 1.1. The ABNF is defined in the proposal as:

```
structure_statement = "structure" ws `identifier` ws [with_clause] shape_members
union_statement = "union" ws `identifier` ws [with_clause] shape_members
with_clause = "with" ws 1*shape_id ws
```

File: designs/mixins.md (hunk 5 above)
Comment (mtdowling) [reply designs/mixins.md:706]: First. I'll clarify.

File: designs/mixins.md (hunk 6 above)
Comment (mtdowling) [reply designs/mixins.md:786]: This section is non-normative, and it's only specific to the reference implementation. Other implementations are free to implement however necessary. This document is not the specification, it's a design document. This is new to the repo, and it's the first one we've published. Once we have agreement here, we'll create a PR to add this to the specification (but also, we're not shipping this until we've finalized 1.1 of the IDL).

File: designs/mixins.md (hunk 4 above)
Comment (mtdowling) [reply designs/mixins.md:369]: There is not. We could allow traits to be filtered per member by extending the `@override` trait. However, I think that needing to _remove_ a trait from a specific member implies a case where the mixin is not appropriate, and the site of the mixin isn't actually opting into the mixin. Adding this kind of filtering per member, I think, would make the models harder to understand (plus harder to implement in Smithy implementations).

An alternative to removing traits but still leveraging mixins is to either not use a mixin for the one-off cases, or to make two layers of mixins. For example, say that some applications of a mixin want a member to be required, and others to be optional. In that case, make two mixins:

```
@mixin
structure FooMixinOptional {
    someMember: String
    otherMember: String
}

@mixin
structure FooMixinRequired with FooMixin {
    @override
    @required
    someMember: String
}
```

**Comment by johnstonskj** [comment]
Awesome, thanks!

File: designs/mixins.md (hunk 5 above)
Comment (mtdowling) [reply designs/mixins.md:706]: Mulled this over with others, and I think we should start with a predefined ordering of members based on what we already had -- mixin members come before members defined locally, mixin members are ordered based on when they are first seen, and redefining members has no effect on ordering. If this proves insufficient for ordering, then we can later consider adding traits for explicit ordering.

**Review by kstich — APPROVED** [review]

**Pull request merged**